packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -937,6 +1183,51 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
     return _currentBottomSheet;
   }
 
+  // FAB API
+  AnimationController _fabMoveController;
+  FabMotionAnimator _fabMotionAnimator;
+  FabPositioner _previousFabPositioner;
+  FabPositioner _fabPositioner;
+  FabPositioner _nextFabPositioner;
+
+  // Moves the FAB to the _nextFabPositioner if the _nextFabPositioner is valid.
+  void _animateFabToNextPositioner() {
+    // If there is no next position to move to, don't animate.
+    if (_nextFabPositioner == null)
+      return;
+    // If the FAB is moving right now, we don't need to start an animation.
+    if (_fabMoveController.isAnimating)
+      return;
+    setState(() {
+      _previousFabPositioner = _fabPositioner;
+      _fabPositioner = _nextFabPositioner;
+      _nextFabPositioner = null;
+    });
+
+    // Only animate if there is a Floating Action Button to animate with.
+    // Otherwise, simply make sure the fab's position is fully updated in 
+    // case it is added.
+    if (widget.floatingActionButton != null) {
+      _fabMoveController.forward(from: 0.0);
+    } else {
+      _fabMoveController.value = _fabMoveController.upperBound;
+    }
+  }
+
+  // Listens to the _fabMoveController and schedules new fab animations if necessary
+  // when the animation completes.
+  void _onFabMoveStatus(AnimationStatus status) {
+    if (status == AnimationStatus.completed) {
+      _animateFabToNextPositioner();
+    }
+  }
+
+  void _moveFab(final FabPositioner newPositioner) {
+    setState(() {
+      _nextFabPositioner = newPositioner;
```

Comment:
this doesn't seem to handle moving the fab while it's moving.
Suppose someone hits a button that moves the fab from left to right, then 80% of the way through the animation they hit "back". we'd want it to start from that 80% point, not from the far right.

Reply:
Getting that will take a bit of extra plumbing:

 - We have to create an animation-dependent positioner tween that can find the correct offset between the old and new positioners.  This requires wiring the positioning logic through the layout phase and supporting an arbitrary level of nesting depth.

 - For scale and rotation tweens, the same concern applies, and we have to decide on a 'bridge' animation to connect the old to the new without any jump cuts.  For most cases, I imagine a Train animation would work.

I am concerned about the offset animation because we need to compute it for each fabPositioner we're moving between at each layout iteration.  If the number of positioners grows beyond the single digits, that could impact performance.  After I finish the other changes, I will explore the instant-response animation to make sure that it works correctly.

Reply:
I haven't studied exactly what it would take to do this, but I do think we definitely need to do this. It hasn't historically been hard for us to do this, e.g. the shape border lerps support this fine (you can lerp from a circle to a rect, then in the middle of that lerp from the half-circle-half-rectangle shape to a circle).